cloud/logstream.go

```diff
@@ -5,69 +5,190 @@ import (
 	"io"
 	"sync/atomic"
 
-	"github.com/earthly/cloud-api/logstream"
-	grpc_retry "github.com/grpc-ecosystem/go-grpc-middleware/retry"
+	pb "github.com/earthly/cloud-api/logstream"
+	"github.com/earthly/earthly/util/stringutil"
+	"github.com/hashicorp/go-multierror"
 	"github.com/pkg/errors"
 	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
 )
 
-// Deltas is a type for iterating over logstream deltas.
-type Deltas interface {
-	Next(ctx context.Context) ([]*logstream.Delta, error)
+func (c *Client) StreamLogs(ctx context.Context, man *pb.RunManifest, ch <-chan *pb.Delta, verbose bool) error {
+	if man.GetResumeToken() == "" {
+		man.ResumeToken = stringutil.RandomAlphanumeric(40)
+	}
+	var (
+		errs  []error
+		retry bool
+		last  *pb.Delta
+	)
+	for {
+		first := []*pb.Delta{firstDelta(man, retry)}
+		if last != nil {
+			first = append(first, last)
+		}
+		var err error
+		last, err = c.streamLogsAttempt(ctx, man.GetBuildId(), first, ch)
+		if err != nil {
+			if retryable(err) {
+				retry = true
+				if verbose {
+					errs = append(errs, err)
```

Comment:
this feels problematic, if the error is retryable, and on the next attempt it succeeds we will still have appended the initial error to `errs`, which will ultimately get returned below under:

```
	if len(errs) > 0 {
		return &multierror.Error{Errors: errs}
	}
```

Reply:
maybe we need to change:

```
if err != nil {
   ...
}
break // change this to a `return nil` instead?
```

Reply:
Well, the idea was to use the `verbose` argument to help surface _all_ errors that were encountered during streaming, even those that were retryable. When `--verbose` is set, we get a nice printout of all of the errors (below). Without the flag, the recoverable error messages are suppressed. 

```
Error(s) in logbus: 1 error occurred:
	* log streamer: 9 errors occurred:
	* rpc error: code = Unavailable desc = something broke
	* rpc error: code = Unavailable desc = something broke
	* rpc error: code = Unavailable desc = something broke
	...
```

Reply:
I like the idea of `--verbose` displaying recovered errors.

I'm trying to understand how a caller would differentiate between a non-recoverable error and an error that was recovered, because simply checking `if err != nil` wouldn't work in this case.

What if instead we pass in a reference to `VerbosePrintf` or some console logger instead -- that way we would display these errors as they occur rather than waiting for all the errors to occur -- this could be especially helpful where we perform a `time.Sleep(...)` between retry attempts.

Reply:
I suppose it would be helpful to differentiate between the 2 error types. I do like the idea of keeping a summary as it's difficult to catch the errors when they're printed inline.

We could use a new error type to capture whether the error was recoverable. In that case, the client can decide how to handle the errors. We could also remove the `verbose` argument from this function. That would be nice since error display isn't really the concern of this component.
```
// Create an error type that specifies whether the error was recoverable.
type StreamError {
  Recoverable bool
  Err err
}
```